### PR TITLE
Remove py27 CI jobs

### PR DESCRIPTION
**Pull request by h-vetinari**
The start of a looooooooong process to get rid of the py2 compat code.

I kept most of the CI jobs that used to be 2.7 at the lower end of the various supported versions, with 2x 2.7->3.5, 1x 2.7->3.6 and only one real upgrade from 2.7->3.7 (first CI job for 3.7 & windows).

@jreback @TomAugspurger

### CODE REVIEW

File: ci/deps/azure-windows-37.yaml [hunk 1]
```diff
@@ -8,14 +8,14 @@ dependencies:
   - dateutil
   - gcsfs
   - html5lib
-  - jinja2=2.8
+  - jinja2
   - lxml
-  - matplotlib=2.0.1
+  - matplotlib=3.0.1
```
Comment (h-vetinari) [inline ci/deps/azure-windows-37.yaml]: Kept fixed MPL pin to have more version coverage (as it was before for 2.0.1). There's a floating MPL version somewhere already.

**Comment by jreback** [comment]
keep in mind that we likely won't merge anything PY2 removal for a month or so.

**Comment by h-vetinari** [comment]
> keep in mind that we likely won't merge anything PY2 removal for a month or so.

Which makes a lot of sense from a backports POV, but I didn't think about that. Guess I just couldn't/can't wait. ;)

**Comment by h-vetinari** [comment]
@TomAugspurger @jorisvandenbossche @jreback 

First passing commit here. Might still need a big of work for the weird SQL/excel failures, but a few PY35-only failures don't seem so critical IMO (dropping 3.5 when 3.8 is released at the end of 2019?).

File: doc/source/whatsnew/v0.25.0.rst [hunk 2]
```diff
@@ -28,7 +28,38 @@ Other Enhancements
 Backwards incompatible API changes
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-- :meth:`Timestamp.strptime` will now rise a NotImplementedError (:issue:`21257`)
```
Comment (h-vetinari) [inline doc/source/whatsnew/v0.25.0.rst]: Fixed a typo here ("rise" -> "raise"), which is why the diff shifts it down

File: doc/source/whatsnew/v0.25.0.rst [hunk 3]
```diff
@@ -28,7 +28,38 @@ Other Enhancements
...
++-----------------+-----------------+----------+
+| xlsxwriter      | 0.7.7           |          |
++-----------------+-----------------+----------+
+| xlwt            | 1.0.0           |          |
++-----------------+-----------------+----------+
+
+.. _whatsnew_0250.api.other:
+
+Other API Changes
+^^^^^^^^^^^^^^^^^
```
Comment (h-vetinari) [inline doc/source/whatsnew/v0.25.0.rst]: This (and the above) reflects the structure of the breaking-changes-section in v.0.24.0

**Review by h-vetinari — COMMENTED** [review]
Looking at the state of `tox.ini` it's obvious that I hasn't run for more than 2 years. Why are we still carrying it around?

File: doc/source/whatsnew/v0.25.0.rst (hunk 2 above)
Comment (h-vetinari) [reply doc/source/whatsnew/v0.25.0.rst]: Actually, the typo was not the only error here. The issue should be #25016, PR was #25124.

Might be worth fixing separately, but I've also fixed it here.

@mroeschke @saurav2608

Comment (h-vetinari) [reply doc/source/whatsnew/v0.25.0.rst]: Actually, the typo was not the only error here. The issue should be #25016, PR was #25124.

Might be worth fixing separately, but I've also fixed it here.

@mroeschke @saurav2608

File: doc/source/whatsnew/v0.25.0.rst [hunk 4]
```diff
@@ -28,7 +28,38 @@ Other Enhancements
...
+
+- :meth:`Timestamp.strptime` will now raise a NotImplementedError (:issue:`25016`)
```
Comment (h-vetinari) [inline doc/source/whatsnew/v0.25.0.rst]: See here.

**Review by h-vetinari — COMMENTED** [review]
@jreback @mroeschke @saurav2608

File: doc/source/whatsnew/v0.25.0.rst (hunk 2 above)
Comment (saurav2608) [reply doc/source/whatsnew/v0.25.0.rst]: @h-vetinari - thanks. I will correct it independent of this.

**Comment by h-vetinari** [comment]
Finally back to green...
Added some more fixes in `install.rst`.

**Comment by jreback** [comment]
this will have to wait for #25554 (which needs to wait for 0.24.2 release)

**Comment by h-vetinari** [comment]
Failure of travis doc job seems to be due to jupyter/notebook#4311.

File: ci/deps/azure-35-compat.yaml [hunk 5]
```diff
@@ -3,26 +3,27 @@ channels:
...
-  - python=2.7*
-  - pytz=2013b
+  - python=3.5*
+  - pytz=2015.4
   - scipy=0.18.1
   - xlrd=1.0.0
-  - xlsxwriter=0.5.2
-  - xlwt=0.7.5
+  - xlsxwriter=0.7.7
+  - xlwt=1.0.0
```
Comment (jreback) [inline ci/deps/azure-35-compat.yaml]: make note of this bump

Comment (jreback) [reply ci/deps/azure-35-compat.yaml]: is this changing our min for testing?

File: ci/deps/azure-36-locale.yaml [hunk 6]
```diff
@@ -11,14 +12,13 @@ dependencies:
...
   - python-dateutil
   - python-blosc
-  - python=2.7
-  - pytz
-  - pytz=2013b
+  - python=3.6
+  - pytz=2016.10
   - scipy
-  - sqlalchemy=0.8.1
+  - sqlalchemy=1.1.4
```
Comment (jreback) [inline ci/deps/azure-36-locale.yaml]: is this changing our min for testing?

File: ci/deps/travis-35.yaml [hunk 7]
```diff
@@ -3,14 +3,13 @@ channels:
   - defaults
   - conda-forge
 dependencies:
+  - backports.lzma
```
Comment (jreback) [inline ci/deps/travis-35.yaml]: is this needed?

File: ci/deps/travis-35.yaml [hunk 8]
```diff
@@ -19,33 +18,40 @@ dependencies:
...
+  - xlwt=1.0.0
   # universal
-  - pytest>=4.0.2
   - pytest-xdist
   - pytest-mock
-  - moto==1.3.4
+  - moto
   - hypothesis>=3.58.0
   - isort
+  # satisfy fastparquet deps with conda
```
Comment (jreback) [inline ci/deps/travis-35.yaml]: these are needed explicity?

File: ci/deps/travis-36-doc.yaml [hunk 9]
```diff
@@ -15,10 +15,10 @@ dependencies:
   - ipywidgets
   - lxml
   - matplotlib
-  - nbconvert
+  - nbconvert>=5.4.1
   - nbformat
   - nbsphinx
-  - notebook
+  - notebook>=5.7.5
```
Comment (jreback) [inline ci/deps/travis-36-doc.yaml]: why are these needed?

File: ci/deps/travis-36-doc.yaml [hunk 10]
```diff
@@ -15,10 +15,10 @@ dependencies:
...
   - numexpr
   - numpy=1.13*
```
Comment (jreback) [inline ci/deps/travis-36-doc.yaml]: just unpin this

File: pandas/tests/test_sorting.py [hunk 11]
```diff
@@ -7,7 +7,7 @@
 from numpy import nan
 import pytest
 
-from pandas.compat import PY2
+from pandas.compat.numpy import _np_version_under1p14
```
Comment (jreback) [inline pandas/tests/test_sorting.py]: remove all actual PY2 changes from this PR, this should only be about changing the CI itself.

**Review by jreback — CHANGES_REQUESTED** [review]
did you edit setup.py as well (near the bottom)?

**Comment by jreback** [comment]
@h-vetinari ok, let's work on getting this in a good state; will merge before the numpy bump. THEN we can change actual code.

**Comment by jreback** [comment]
eg.
```
setup(name=DISTNAME,
      maintainer=AUTHOR,
      version=versioneer.get_version(),
      packages=find_packages(include=['pandas', 'pandas.*']),
      package_data={'': ['templates/*', '_libs/*.dll']},
      ext_modules=maybe_cythonize(extensions, compiler_directives=directives),
      maintainer_email=EMAIL,
      description=DESCRIPTION,
      license=LICENSE,
      cmdclass=cmdclass,
      url=URL,
      download_url=DOWNLOAD_URL,
      long_description=LONG_DESCRIPTION,
      classifiers=CLASSIFIERS,
      platforms='any',
      python_requires='>=2.7,!=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*,!=3.4.*',
      **setuptools_kwargs)
```

**Comment by TomAugspurger** [comment]
Matplotlib includes this: https://github.com/matplotlib/matplotlib/blob/f9a63d77b7d17afbc7eba4c30e12e5e0888c5115/setup.py#L6-L21 as a check, for older versions of pip / setuptools.

No preference on whether we do that here or elsewhere.

**Comment by jreback** [comment]
@TomAugspurger what I meant was it needs to be updated, e.g. 

``python_requires='>=3.5.*` i think might work

**Comment by TomAugspurger** [comment]
Right. It's also recommended to include something like MPL's code for older
versions of pip / setuptools that don't honor python_requires:
https://python3statement.org/practicalities/

File: ci/deps/travis-36-doc.yaml (hunk 10 above)
Comment (h-vetinari) [reply ci/deps/travis-36-doc.yaml]: I'm doing that already in #25554 but OK

File: pandas/tests/test_sorting.py (hunk 11 above)
Comment (h-vetinari) [reply pandas/tests/test_sorting.py]: Of course I'm not aiming to change code here, but `test_unsortable` below would fail otherwise. The current regex is not general enough (the skipif for `PY2` is a cop-out IMO), and *will* need to be fixed for a passing CI.

**Review by h-vetinari — COMMENTED** [review]
@jreback 
Not sure what you mean with your [comment](https://github.com/pandas-dev/pandas/pull/24942#issuecomment-472833902).

Can we remove `tox.ini` - it's been unused for >2 years.

File: ci/deps/travis-35.yaml (hunk 7 above)
Comment (jreback) [reply ci/deps/travis-35.yaml]: yes

File: ci/deps/travis-35.yaml [hunk 12]
```diff
@@ -19,33 +18,40 @@ dependencies:
   - numexpr
   - numpy=1.13*
   - openpyxl=2.4.0
+  - pathlib
```
Comment (jreback) [inline ci/deps/travis-35.yaml]: this is not needed as its builtin

File: ci/deps/travis-35.yaml (hunk 8 above)
Comment (jreback) [reply ci/deps/travis-35.yaml]: i would rather remove fastparquet from this build then, we have it installed for 1 or more of 3.6/3.7 already I think (pls check)

File: pandas/tests/io/test_excel.py [hunk 13]
```diff
@@ -2182,6 +2182,8 @@ def test_write_cells_merge_styled(self, merge_cells, ext, engine):
             assert xcell_b1.font == openpyxl_sty_merged
             assert xcell_a2.font == openpyxl_sty_merged
 
+    @pytest.mark.xfail(PY35 and not PY36, reason='only fails on Linux?',
```
Comment (jreback) [inline pandas/tests/io/test_excel.py]: why is this changing?

File: pandas/tests/io/test_sql.py [hunk 14]
```diff
@@ -28,7 +28,7 @@
 import pytest
 
 import pandas.compat as compat
-from pandas.compat import PY2, PY36, lrange, range, string_types
+from pandas.compat import PY2, PY35, PY36, lrange, range, string_types
```
Comment (jreback) [inline pandas/tests/io/test_sql.py]: really really don't change things, if you are forcing chanegs then roll back the env changes which cause this.

File: pandas/tests/test_algos.py [hunk 15]
```diff
@@ -11,8 +11,9 @@
 
 from pandas._libs import (
     algos as libalgos, groupby as libgroupby, hashtable as ht)
-from pandas.compat import PY2, lrange, range
-from pandas.compat.numpy import np_array_datetime64_compat
+from pandas.compat import lrange, range
```
Comment (jreback) [inline pandas/tests/test_algos.py]: see my comment above

**Review by jreback — CHANGES_REQUESTED** [review]
@h-vetinari this is again changing too much. I want *only* the actual env files to be changed. you are changing all kinds of CI things and something is still breaking. I would start anew and simply remove the PY2 builds and leave the actual updates of versions to another PR.

File: pandas/tests/io/test_excel.py (hunk 13 above)
Comment (h-vetinari) [reply pandas/tests/io/test_excel.py]: Because I wasn't able to make this test work on PY35 regardless of the versions I tried.

File: pandas/tests/test_algos.py (hunk 15 above)
Comment (h-vetinari) [reply pandas/tests/test_algos.py]: Again, if I don't fix the regex here, the tests will fail.

File: ci/deps/travis-35.yaml (hunk 8 above)
Comment (h-vetinari) [reply ci/deps/travis-35.yaml]: Removed. Remaining jobs with `fastparquet`:
```
ci/deps\azure-windows-36.yaml:9:  - fastparquet>=0.2.1
ci/deps\travis-36-doc.yaml:9:  - fastparquet>=0.2.1
ci/deps\travis-36.yaml:10:  - fastparquet>=0.2.1
```

**Review by h-vetinari — COMMENTED** [review]
@jreback 
Pushing some changes based on your review. This PR contains some xfails that I wasn't able to solve so far, as well as two changes to regexes that would otherwise fail. If the goal is a passing CI, those are unavoidable.

**Comment by jorisvandenbossche** [comment]
@h-vetinari can you point to the test errors for which you needed to add the additional xfails? (or do a new build without any of them) Ideally, we should investigate each of those in more detail why they are failing instead of just xfailing them (unless you already know the reason for each of them)

**Comment by h-vetinari** [comment]
@jorisvandenbossche 
The fails are distributed over the CI runs throughout the thread, but I've reverted them in the last commit (including the regex fixes @jreback wanted me to revert), so then there's a single place to inspect the failures. Also - who knows - maybe some of them resolved themselves...

**Comment by h-vetinari** [comment]
@jorisvandenbossche @TomAugspurger @jreback 
Please check out https://travis-ci.org/pandas-dev/pandas/jobs/506535600, it fails already in the `single` stage with the tests I would have xfailed. When I first started this PR I played around with bumping the versions of xlrd etc (even when unpinning `pymysql/xlsxwriter`), but that didn't solve things either.

Considering the fact that removing support for PY35 is already being talked about in #25725, I think a small handful of xfailed tests is not a big burden. Certainly those are less trivial to fix than for example the skips that have been added for regex fails (of which there about 30).

**Comment by jreback** [comment]
as i said above pls start over and JUST remove the 2.7 jobs; since things are working now this will just work. after that you can do independent PR to add new / change things.

something is breaking and certainly needs fixing
but not here

**Comment by jreback** [comment]
PY35 is not being removed at this time

**Comment by h-vetinari** [comment]
> @jreback: as i said above pls start over and JUST remove the 2.7 jobs; since things are working now this will just work. after that you can do independent PR to add new / change things.

Of course it's easier to just rip things out, but I actually went to the effort here of thinking of a good distribution for the CI, updating the jobs, and making them pass again.

That's also how the last numpy bump was made. Tbh, I feel these conditions are quite arbitrary, but well... I'll delete the former PY2 jobs later today, and post the fixes in a separate PR.

**Comment by jreback** [comment]
@h-vetinari as I have explained many many times. PRs which do lots of things are take *way* more time to review and are much harder to merge. It is the same case here.

**Comment by h-vetinari** [comment]
@jreback 
This is green now. If only it always were as easy as deleting CI jobs, haha.

File: doc/source/whatsnew/v0.25.0.rst [hunk 16]
```diff
@@ -70,12 +70,27 @@ is respected in indexing. (:issue:`24076`, :issue:`16785`)
...
+Due to dropping support for Python 2.7, a number of optional dependencies have updated minimum versions.
+Independently, some minimum supported versions of dependencies were updated (:issue:`23519`, :issue:`24942`).
 If installed, we now require:
 
 +-----------------+-----------------+----------+
 | Package         | Minimum Version | Required |
 +=================+=================+==========+
+| beautifulsoup4  | 4.4.1           |          |
++-----------------+-----------------+----------+
+| openpyxl        | 2.2.6           |          |
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: are these the minimums implied by the py3 builds?

**Review by jreback — CHANGES_REQUESTED** [review]
just the 1 comment

File: doc/source/whatsnew/v0.25.0.rst (hunk 16 above)
Comment (jreback) [reply doc/source/whatsnew/v0.25.0.rst]: @h-vetinari ?

Comment (h-vetinari) [reply doc/source/whatsnew/v0.25.0.rst]: Since you asked me to remove the compat jobs, all of the libraries here are unpinned for the rest of the remaining CI. They are however the minimum of what conda will resolve for PY35, so these are the implied minimums as you say.

I will re-add another PR with all the work that was done here on the PY35 jobs anyway.

**Review by h-vetinari — COMMENTED** [review]
Sorry, somehow didn't get a notification for the first comment.

File: doc/source/whatsnew/v0.25.0.rst (hunk 16 above)
Comment (jreback) [reply doc/source/whatsnew/v0.25.0.rst]: ok thanks. pls do so.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks, would also appreciate a PR against : https://github.com/Macpython/pandas-wheels (the daily branch, not master).

**Comment by h-vetinari** [comment]
> thanks, would also appreciate a PR against : https://github.com/Macpython/pandas-wheels (the daily branch, not master).

This same PR or the PY35 one?

**Comment by jreback** [comment]
neither. one to remove the PY2 jobs only (this is where we build wheels), and is our daily testing.
